File: crystal-gui/src/llm_client.rs
//! LLM Client for communicating with Ollama
//!
//! Provides chat interface to local LLM models via Ollama API with native tool calling.

use serde::{Deserialize, Serialize};
use serde_json::Value;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum LlmError {
    #[error("HTTP request failed: {0}")]
    RequestError(#[from] reqwest::Error),
    #[error("Ollama not available at {0}")]
    NotAvailable(String),
    #[error("Model not found: {0}")]
    ModelNotFound(String),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatMessage {
    pub role: String,
    pub content: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_calls: Option<Vec<ToolCall>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolCall {
    pub function: ToolFunction,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolFunction {
    pub name: String,
    pub arguments: Value,
}

#[derive(Debug, Clone, Serialize)]
struct OllamaToolDef {
    #[serde(rename = "type")]
    tool_type: String,
    function: OllamaFunctionDef,
}

#[derive(Debug, Clone, Serialize)]
struct OllamaFunctionDef {
    name: String,
    description: String,
    parameters: Value,
}

#[derive(Debug, Clone, Serialize)]
struct OllamaChatRequest {
    model: String,
    messages: Vec<ChatMessage>,
    stream: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    tools: Option<Vec<OllamaToolDef>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    format: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
struct OllamaChatResponse {
    message: ResponseMessage,
}

#[derive(Debug, Clone, Deserialize)]
struct ResponseMessage {
    #[serde(default)]
    content: String,
    #[serde(default)]
    tool_calls: Option<Vec<ToolCall>>,
}

#[derive(Debug, Clone, Deserialize)]
struct OllamaTagsResponse {
    models: Vec<OllamaModel>,
}

#[derive(Debug, Clone, Deserialize)]
struct OllamaModel {
    name: String,
}

/// Result from chat_with_tools - either tool calls or plain text
pub enum ChatResult {
    ToolCalls(Vec<ToolCall>),
    Text(String),
}

pub struct LlmClient {
    base_url: String,
    model: String,
    client: reqwest::blocking::Client,
    available_models: Vec<String>,
}

impl LlmClient {
    pub fn new(base_url: &str, model: &str) -> Self {
        Self {
            base_url: base_url.to_string(),
            model: model.to_string(),
            client: reqwest::blocking::Client::new(),
            available_models: Vec::new(),
        }
    }

    pub fn check_connection(&mut self) -> Result<bool, LlmError> {
        let url = format!("{}/api/tags", self.base_url);
        let response = self.client.get(&url).send()?;

        if response.status().is_success() {
            let tags: OllamaTagsResponse = response.json()?;
            self.available_models = tags.models.into_iter().map(|m| m.name).collect();
            Ok(true)
        } else {
            Err(LlmError::NotAvailable(self.base_url.clone()))
        }
    }

    pub fn get_available_models(&self) -> &[String] {
        &self.available_models
    }

    pub fn set_model(&mut self, model: &str) {
        self.model = model.to_string();
    }

    pub fn get_model(&self) -> &str {
        &self.model
    }

    /// Chat with native Ollama tool calling
    /// 
    /// tools_json: Array of MCP tool definitions (from get_tools)
    /// Returns either tool calls from the LLM or plain text
    pub fn chat_with_tools(
        &self,
        messages: &[ChatMessage],
        tools_json: &[Value],
    ) -> Result<ChatResult, LlmError> {
        let url = format!("{}/api/chat", self.base_url);
        
        // Convert MCP tools to Ollama format - NO FILTERING, send all tools
        let ollama_tools: Vec<OllamaToolDef> = tools_json
            .iter()
            .filter_map(|tool| {
                let name = tool.get("name")?.as_str()?.to_string();
                let description = tool
                    .get("description")
                    .and_then(|d| d.as_str())
                    .unwrap_or("")
                    .to_string();
                let parameters = tool
                    .get("inputSchema")
                    .cloned()
                    .unwrap_or(serde_json::json!({"type": "object", "properties": {}}));

                Some(OllamaToolDef {
                    tool_type: "function".to_string(),
                    function: OllamaFunctionDef {
                        name,
                        description,
                        parameters,
                    },
                })
            })
            .collect();

        // Use messages as-is - no hardcoded system prompts
        // Tool schemas are sent via the tools parameter (standard MCP/OpenAI format)
        let request = OllamaChatRequest {
            model: self.model.clone(),
            messages: messages.to_vec(),
            stream: false,
            tools: if ollama_tools.is_empty() {
                None
            } else {
                Some(ollama_tools)
            },
            // DO NOT use format: "json" - it DISABLES native tool calling!
            // Native tool calling returns tool_calls array, which is the proper MCP-compatible approach
            format: None,
        };
        
        // DEBUG: Print exact LLM request including tool schemas
        eprintln!("\n[DEBUG] ========== LLM REQUEST ==========");
        eprintln!("[DEBUG] Model: {}", request.model);
        eprintln!("[DEBUG] Messages ({}):", request.messages.len());
        for (i, msg) in request.messages.iter().enumerate() {
            eprintln!("[DEBUG]   [{}] {}: {}", i, msg.role, &msg.content[..msg.content.len().min(200)]);
        }
        if let Some(ref tools) = request.tools {
            eprintln!("[DEBUG] Tools ({}) with FULL SCHEMAS:", tools.len());
            for t in tools {
                eprintln!("[DEBUG]   - {} ({})", t.function.name, t.function.description.chars().take(60).collect::<String>());
                // Show schema parameters
                if let Some(props) = t.function.parameters.get("properties") {
                    let keys: Vec<_> = props.as_object().map(|o| o.keys().cloned().collect()).unwrap_or_default();
                    eprintln!("[DEBUG]     Parameters: {:?}", keys);
                }
            }
        }
        eprintln!("[DEBUG] Format: {:?} (None = native tool calling)", request.format);
        eprintln!("[DEBUG] =====================================\n");

        let response = self.client.post(&url).json(&request).send()?;

        if response.status().is_success() {
            let chat_response: OllamaChatResponse = response.json()?;

            // Check for tool calls first
            if let Some(tool_calls) = chat_response.message.tool_calls {
                if !tool_calls.is_empty() {
                    return Ok(ChatResult::ToolCalls(tool_calls));
                }
            }

            // Fall back to text content
            Ok(ChatResult::Text(chat_response.message.content))
        } else {
            Err(LlmError::ModelNotFound(self.model.clone()))
        }
    }

    /// Legacy method for plain chat without tools (for backward compatibility)
    pub fn chat(&self, messages: &[ChatMessage], format: Option<String>) -> Result<String, LlmError> {
        let url = format!("{}/api/chat", self.base_url);

        let request = OllamaChatRequest {
            model: self.model.clone(),
            messages: messages.to_vec(),
            stream: false,
            tools: None,
            format,
        };

        let response = self.client.post(&url).json(&request).send()?;

        if response.status().is_success() {
            let chat_response: OllamaChatResponse = response.json()?;
            Ok(chat_response.message.content)
        } else {
            Err(LlmError::ModelNotFound(self.model.clone()))
        }
    }
}

impl Default for LlmClient {
    fn default() -> Self {
        Self::new("http://localhost:11434", "qwen2.5:7b")
    }
}

File: crystal-gui-web/src/services/llmClient.ts
/**
 * LLM Client Service - Ollama API communication
 * 
 * Ported from: crystal-gui/src/llm_client.rs
 * Communicates with local Ollama server for LLM inference
 * 
 * DEBUG: Comprehensive logging enabled
 */

import type { ChatResult, McpTool, ToolCall } from '../types';
import { debug, debugError, debugTimer, debugInspect } from '../debug';

const DEFAULT_OLLAMA_URL = 'http://localhost:11434';
const DEFAULT_MODEL = 'qwen2.5:7b';

interface OllamaToolDef {
    type: 'function';
    function: {
        name: string;
        description: string;
        parameters: object;
    };
}

interface OllamaChatRequest {
    model: string;
    messages: Array<{ role: string; content: string }>;
    stream: boolean;
    tools?: OllamaToolDef[];
    format?: string;
    options?: {
        temperature?: number;
        top_p?: number;
        seed?: number;
    };
}

interface OllamaChatResponse {
    model: string;
    message: {
        role: string;
        content: string;
        tool_calls?: ToolCall[];
    };
    done: boolean;
}

/**
 * Convert MCP tools to Ollama format
 */
function mcpToolsToOllamaFormat(tools: McpTool[]): OllamaToolDef[] {
    debug('LLM_CLIENT', `Converting ${tools.length} MCP tools to Ollama format`);
    return tools.map(tool => ({
        type: 'function',
        function: {
            name: tool.name,
            description: tool.description || '',
            parameters: tool.inputSchema || { type: 'object', properties: {} },
        },
    }));
}

/**
 * LLM Client for communicating with Ollama
 */
export class LlmClient {
    private baseUrl: string;
    private model: string;
    private availableModels: string[] = [];
    private _temperature: number = 0; // Default 0 for deterministic tool calling

    constructor(baseUrl: string = DEFAULT_OLLAMA_URL, model: string = DEFAULT_MODEL) {
        this.baseUrl = baseUrl;
        this.model = model;
        debug('LLM_CLIENT', `Created LLM client: url=${baseUrl}, model=${model}, temperature=${this._temperature}`);
    }

    /**
     * Check connection to Ollama and fetch available models
     */
    async checkConnection(): Promise<boolean> {
        const endTimer = debugTimer('LLM_CLIENT', 'checkConnection');
        debug('LLM_CLIENT', `Checking connection to: ${this.baseUrl}/api/tags`);

        try {
            const response = await fetch(`${this.baseUrl}/api/tags`);
            debug('LLM_CLIENT', `Response status: ${response.status}`);

            if (response.ok) {
                const data = await response.json();
                this.availableModels = (data.models || []).map((m: { name: string }) => m.name);
                debug('LLM_CLIENT', `✅ Connected. Available models: ${this.availableModels.join(', ')}`);
                endTimer();
                return true;
            }
            debug('LLM_CLIENT', '❌ Connection failed: not OK response');
            endTimer();
            return false;
        } catch (error) {
            debugError('LLM_CLIENT', error, 'Connection check failed');
            endTimer();
            return false;
        }
    }

    /**
     * Get available models
     */
    getAvailableModels(): string[] {
        return this.availableModels;
    }

    /**
     * Set model to use
     */
    setModel(model: string): void {
        debug('LLM_CLIENT', `Changing model from ${this.model} to ${model}`);
        this.model = model;
    }

    /**
     * Get current model
     */
    getModel(): string {
        return this.model;
    }

    /**
     * Set temperature (0-1). Lower = more deterministic, higher = more creative.
     * Default 0 for precise tool calling.
     */
    setTemperature(temp: number): void {
        this._temperature = Math.max(0, Math.min(1, temp));
        debug('LLM_CLIENT', `Temperature set to ${this._temperature}`);
    }

    /**
     * Get current temperature
     */
    getTemperature(): number {
        return this._temperature;
    }

    /**
     * Chat with native Ollama tool calling
     * 
     * @param messages - Chat message history
     * @param tools - MCP tools to make available
     * @returns ChatResult - either tool calls or text response
     */
    async chatWithTools(
        messages: Array<{ role: string; content: string }>,
        tools: McpTool[]
    ): Promise<ChatResult> {
        const endTimer = debugTimer('LLM_CLIENT', 'chatWithTools');

        debug('LLM_CLIENT', '='.repeat(60));
        debug('LLM_CLIENT', '📤 STARTING LLM CHAT WITH TOOLS');
        debug('LLM_CLIENT', `Model: ${this.model}`);
        debug('LLM_CLIENT', `Message count: ${messages.length}`);
        debug('LLM_CLIENT', `Tool count: ${tools.length}`);
        debug('LLM_CLIENT', `Last message: "${messages[messages.length - 1]?.content?.slice(0, 100)}..."`);

        const ollamaTools = mcpToolsToOllamaFormat(tools);
        debug('LLM_CLIENT', `Converted tools: ${ollamaTools.map(t => t.function.name).join(', ')}`);

        const request: OllamaChatRequest = {
            model: this.model,
            messages: messages,
            stream: false,
            tools: ollamaTools.length > 0 ? ollamaTools : undefined,
            // DO NOT use format: 'json' - it DISABLES native tool calling!
            format: undefined,
            options: {
                temperature: this._temperature,
            },
        };

        debug('LLM_CLIENT', `Full request:`, debugInspect(request, 2000));

        try {
            debug('LLM_CLIENT', `⏳ Sending POST to ${this.baseUrl}/api/chat...`);
            const startTime = performance.now();

            const response = await fetch(`${this.baseUrl}/api/chat`, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify(request),
            });

            const elapsed = ((performance.now() - startTime) / 1000).toFixed(2);
            debug('LLM_CLIENT', `📥 Response received in ${elapsed}s, status: ${response.status}`);

            if (!response.ok) {
                const errorText = await response.text();
                debugError('LLM_CLIENT', errorText, 'LLM request failed');
                throw new Error(`LLM request failed: ${response.statusText} - ${errorText}`);
            }

            const data: OllamaChatResponse = await response.json();
            debug('LLM_CLIENT', `Response model: ${data.model}`);
            debug('LLM_CLIENT', `Response done: ${data.done}`);
            debug('LLM_CLIENT', `Message role: ${data.message.role}`);
            debug('LLM_CLIENT', `Message content length: ${data.message.content?.length || 0}`);
            debug('LLM_CLIENT', `Tool calls present: ${!!data.message.tool_calls}`);

            if (data.message.tool_calls) {
                debug('LLM_CLIENT', `Tool calls count: ${data.message.tool_calls.length}`);
            }

            // Check for tool calls first
            if (data.message.tool_calls && data.message.tool_calls.length > 0) {
                debug('LLM_CLIENT', '✅ LLM returned TOOL CALLS:');
                data.message.tool_calls.forEach((tc, i) => {
                    debug('LLM_CLIENT', `  [${i}] ${tc.function.name}(${debugInspect(tc.function.arguments)})`);
                });
                endTimer();
                return {
                    type: 'tool_calls',
                    toolCalls: data.message.tool_calls,
                };
            }

            // Fall back to text content
            debug('LLM_CLIENT', '📝 LLM returned TEXT response:');
            debug('LLM_CLIENT', `Content preview: ${data.message.content?.slice(0, 200)}...`);
            endTimer();
            return {
                type: 'text',
                content: data.message.content,
            };
        } catch (error) {
            debugError('LLM_CLIENT', error, 'chatWithTools exception');
            debug('LLM_CLIENT', '='.repeat(60));
            endTimer();
            throw error;
        }
    }

    /**
     * Plain chat without tools (legacy method)
     */
    async chat(
        messages: Array<{ role: string; content: string }>,
        format?: string
    ): Promise<string> {
        const endTimer = debugTimer('LLM_CLIENT', 'chat (plain)');
        debug('LLM_CLIENT', `Plain chat with ${messages.length} messages`);

        const request: OllamaChatRequest = {
            model: this.model,
            messages,
            stream: false,
            format,
            options: {
                temperature: this._temperature,
            },
        };

        try {
            const response = await fetch(`${this.baseUrl}/api/chat`, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify(request),
            });

            debug('LLM_CLIENT', `Response status: ${response.status}`);

            if (!response.ok) {
                const errorText = await response.text();
                debugError('LLM_CLIENT', errorText, 'Plain chat failed');
                throw new Error(`LLM request failed: ${response.statusText}`);
            }

            const data: OllamaChatResponse = await response.json();
            debug('LLM_CLIENT', `Response content length: ${data.message.content?.length}`);
            endTimer();
            return data.message.content;
        } catch (error) {
            debugError('LLM_CLIENT', error, 'Plain chat exception');
            endTimer();
            throw error;
        }
    }
}

// Singleton instance
export const llmClient = new LlmClient();
